fix(lat_lon_controller): change reference to the node

TauTheLepton opened the pull request:
## Description

This PR is related to issue #3309.
Pure pursuit controller held its parent Node in its own shared_ptr (so the Node object was held by two independent shared_ptrs) which led to dependency issues where shared_ptr in pure_pursuit tried to free node memory which has already been freed. 

I changed pure pursuit controller to hold Node by reference as controller should never outlive its parent node.

I also changed mpc and pid controllers to hold Node by reference instead of a raw pointer. This way it is a bit safer, as reference can not be reassigned.

## Related links

<!-- Write the links related to this PR. Private links should be clearly marked as private, for example, '[FOO COMPANY INTERNAL LINK](https://example.com)'. -->
None

## Tests performed

<!-- Describe how you have tested this PR. -->
I merged these changes into current main and verified that control container finishes cleanly with both mpc and pure pursuit controllers.

As the issue this PR is related to originated from other PR (#3287) I also merged changes from that PR (which is accepted but not yet merged) and all tests (which now includes testing both types of controllers independently) passed.

## Notes for reviewers

<!-- Write additional information if necessary. It should be written if there are related PRs that should be merged at the same time. -->
**IMPORTANT** In the changes I made in order for the smoke tests to pass I deleted [one wait loop](https://github.com/TauTheLepton/autoware.universe/blob/7696f5faf5abb304535356143e059af9be0c295c/control/pure_pursuit/src/pure_pursuit/pure_pursuit_lateral_controller.cpp#L100).

I don't know what its job was and whether it had any job at all. Before merging I am asking someone who knows what was the original purpose of this wait loop to verify whether it can be safely deleted.

From the tests I performed it seems that this deletion does not affect system performance.

## Interface changes

<!-- Describe any changed interfaces, such as topics, services, or parameters. -->
None

## Effects on system behavior

When closing Autoware control container finishes without errors and all tests pass.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

takayuki5168 commented on control/mpc_lateral_controller/src/mpc_lateral_controller.cpp:
```diff
@@ -42,45 +42,44 @@ void update_param(
 }  // namespace
 
-MpcLateralController::MpcLateralController(rclcpp::Node & node) : node_{&node}
+MpcLateralController::MpcLateralController(rclcpp::Node & node) : node_{node}
```
With this change, `node_` is copied from `node`, right?
I think these should be the same variable.

takayuki5168 replied:
Sorry, node_ is a reference as you said.

takayuki5168 commented:
@TauTheLepton 
Thank you for the PR.

I think the problem is that mpc has the node as a raw pointer but pure_pursuit has the node as a shared_ptr.
Then, if a raw pointer is freed, the shared_ptr's target is null and the node will die.

Your change is mpc will have a copied node and pure_pursuit will have another copied node, which I don't think is good since in some cases mpc and pure_pursuit may want to access to the same information via the same node variable.

How about changing mpc having a node as a shared_ptr in the same way as pure_pursuit.
Then mpc and pure_pursit's shared_ptrs will point to the same node variable. Also the problem of freeing will be resolved thanks to shared_ptr's freeing logic I guess.

TauTheLepton commented:
@takayuki5168 

> Your change is mpc will have a copied node and pure_pursuit will have another copied node, which I don't think is good since in some cases mpc and pure_pursuit may want to access to the same information via the same node variable.

In my changes [here](https://github.com/autowarefoundation/autoware.universe/pull/3696/files#diff-8a89e4298b2125992eb5c99ef8f8d12caaf6d9d25646c2e63eee66052fe17a16R71) I declared a `node_` as a reference to the node, so it is almost an equivalent to a raw pointer, but safer as it can't be reassigned.

As to the part that both controllers may want to access the same information it actually is possible with references, as both controllers will just hold a reference to the same node object. Although such case is not possible with current implementation, as both types of controllers are assigned to the save field in the node [here](https://github.com/autowarefoundation/autoware.universe/blob/261ce2139073ad4c61bcbbe897f2d5389c87611a/control/trajectory_follower_node/src/controller_node.cpp#L39-L50)

> How about changing mpc having a node as a shared_ptr in the same way as pure_pursuit. Then mpc and pure_pursit's shared_ptrs will point to the same node variable. Also the problem of freeing will be resolved thanks to shared_ptr's freeing logic I guess.

This is not possible, because in pure_pursuit (and in mpc if we were to make the proposed changes) there is created a _new_ shared_ptr from the memory address of the node. The reference count is not working as you would like it to.

This was the issue. Two shared_ptr were controlling the same node object. This was what was happening upon closing the system:
1. component container called delete on node
2. node called delete on pure_pursuit
3. pure_pursuit called delete on _the same_ node
4. after deleting pure_pursuit, node came back to finish deleting itself, but it was not there, as pure_pursuit has already deleted it and this caused the freeing memory that has already been freed error.

takayuki5168 commented:
@TauTheLepton 
Thank you for the clear explanation.

My concern is managing the node with a raw pointer and a reference since when the pointer is freed, the reference's target will disappear.
Here is a simple example.
https://wandbox.org/permlink/oW2WMGmQbEOizh1W

I understand the same concern exists when managing the node by a raw pointer and a shared_ptr.
Don't you have the same concern?


What do you think about not having a node as a member variable?
This is an example of not having a node as a member variable.
https://github.com/autowarefoundation/autoware.universe/blob/ffc8b712c5bb440b5a19dbd9fe818b1d087fdde8/planning/obstacle_avoidance_planner/src/mpt_optimizer.cpp#L188
https://github.com/autowarefoundation/autoware.universe/blob/ffc8b712c5bb440b5a19dbd9fe818b1d087fdde8/planning/obstacle_avoidance_planner/src/mpt_optimizer.cpp#L421

TauTheLepton commented:
@takayuki5168 
> My concern is managing the node with a raw pointer and a reference since when the pointer is freed, the reference's target will disappear. Here is a simple example. https://wandbox.org/permlink/oW2WMGmQbEOizh1W

Yes, I am aware this is a potential issue. Please see my reasoning below.

> I understand the same concern exists when managing the node by a raw pointer and a shared_ptr. Don't you have the same concern?

I do have the same concerns, but I don't see other solution that retains current features (shown below).

> What do you think about not having a node as a member variable? This is an example of not having a node as a member variable.
> 
> https://github.com/autowarefoundation/autoware.universe/blob/ffc8b712c5bb440b5a19dbd9fe818b1d087fdde8/planning/obstacle_avoidance_planner/src/mpt_optimizer.cpp#L188
> 
> https://github.com/autowarefoundation/autoware.universe/blob/ffc8b712c5bb440b5a19dbd9fe818b1d087fdde8/planning/obstacle_avoidance_planner/src/mpt_optimizer.cpp#L421

Unfortunately I don't think this solution would work here. In the pure_pursuit controller we use other Node functionality like here

https://github.com/autowarefoundation/autoware.universe/blob/140866a1bd7675429bc49d516809136d01144707/control/pure_pursuit/src/pure_pursuit/pure_pursuit_lateral_controller.cpp#L131

Or here

https://github.com/autowarefoundation/autoware.universe/blob/140866a1bd7675429bc49d516809136d01144707/control/pure_pursuit/src/pure_pursuit/pure_pursuit_lateral_controller.cpp#L307

(I post lines from current main, because they generate preview, but the only difference is de-referencing)

> If this idea does not look good to you, I respect your idea.

The reason I don't think this is as big of an issue is because the Node _creates_ a pure_pursuit and never shares it with anybody, so in the current implementation the lifetime of the pure_pursuit is guaranteed to be not longer than the lifetime of the Node.

This could potentially be an issue if Node shared pure_pursuit. To address this potential issue we could make the Node hold both controllers in unique_ptr instead of shared_ptr here.

https://github.com/autowarefoundation/autoware.universe/blob/140866a1bd7675429bc49d516809136d01144707/control/trajectory_follower_node/include/trajectory_follower_node/controller_node.hpp#L69-L70

This would make it a bit safer, because someone would have to actually _move_ the longitudinal controller pointer instead of just copying it.

However I'm not sure if such a change is necessary, let me know if my reasoning addressed your issues and whether you think the change of pointer type is a good idea.

takayuki5168 commented:
@TauTheLepton Sorry to be late. :pray: 

I'd like to suggest another idea.
How about creating a shared_ptr in the controller node and passing it to each lateral and longitudinal controller so that a reference count will work.

Due to the maintainability and readability, I'm looking for a better idea without using a node reference.

TauTheLepton commented:
@takayuki5168 I am very sorry to be late and for the mess with review requests :pray: 

> I'd like to suggest another idea. How about creating a shared_ptr in the controller node and passing it to each lateral and longitudinal controller so that a reference count will work.

I see your point and I would like to make it the way you suggested, but unfortunately it is not possible. The Controller class is used in the [Component conatiner](https://docs.ros.org/en/humble/Concepts/Intermediate/About-Composition.html) - for this reason it is compiled into a library here:
https://github.com/autowarefoundation/autoware.universe/blob/b86d3d8bc59502fbfa5b4bd14e30e0550bf939e0/control/trajectory_follower_node/src/controller_node.cpp#L256-L257

Upon launch the Component Container loads the Controller class into a shared_ptr. The only possibility to use shared_ptr and reference counting correctly is to call a `shared_from_this()` method to obtain the shared_ptr created in the Component Container. Unfortunately this method can be called only after the Controller is initialized, so after the constructor has finished executing.
This is not a good solution as we need the Controller Node for the lateral and longitudinal controller initialization. One possibility would be to create some `init()` method, put all initialization code in it and call it once in every callback, but this would result in very bad code readability.

> Due to the maintainability and readability, I'm looking for a better idea without using a node reference.

I understand and that is why I came up with other solution. Instead of holding some reference to the Controller Node for a whole lifetime of the lateral and longitudinal controller we can use the Controller Node at the initialization of every aforementioned controller for parameter declarations etc.
Then we take only the components of the Controller Node that are necessary for the lateral and longitudinal controller functionality - like clock and logger - and use them instead of the whole Node Controller.

Let me know what you think of this change.

TauTheLepton commented:
I tested this solution by launching `planning_simulator` which worked well and by running `colcon test` which - after applying the changes from #4118 - passed.

takayuki5168 commented:
@TauTheLepton 
Thank you for the much better idea and its code change! I agree with it.
Could you add a comment so that other developers will know the reason to use the node argument as a reference?

I really appreciate your thorough fix again.

takayuki5168 commented:
@TakaHoribe Could you approve the PR please.

TauTheLepton commented:
@takayuki5168 

> @TauTheLepton Thank you for the much better idea and its code change! I agree with it. Could you add a comment so that other developers will know the reason to use the node argument as a reference?

I have added appropriate comments in the header files.

TakaHoribe commented:
@takayuki5168 @TauTheLepton 

Sorry for the late response. I've gone through the thread. Let me confirm if my understanding is correct.

- Two different shared_ptrs are pointing to the same node, causing a double-free issue when the destructor is called.
- Simply replacing the "shared_ptr" with a "raw pointer" can address this issue.
    - From a reassignment perspective, using a "reference" might be safer.
- However, in this case, having a reference or pointer to the parent node could cause a problem if someone deletes the parent node. Therefore, the proposed solution here is to remove the node from the member variables and pass it through args in each function instead. (The parent node will not be deleted by anyone in this case though)

TauTheLepton commented:
@TakaHoribe 

> @takayuki5168 @TauTheLepton
> 
> Sorry for the late response. I've gone through the thread. Let me confirm if my understanding is correct.
> 
> - Two different shared_ptrs are pointing to the same node, causing a double-free issue when the destructor is called.

Yes, this was the case in the `pure_pursuit` controller implementation.
> 
> - Simply replacing the "shared_ptr" with a "raw pointer" can address this issue.
>       
>   - From a reassignment perspective, using a "reference" might be safer.

Correct, the "raw pointer" was used in the original `mpc` controller implementation (which did not produce any errors).
> 
> - However, in this case, having a reference or pointer to the parent node could cause a problem if someone deletes the parent node. Therefore, the proposed solution here is to remove the node from the member variables and pass it through args in each function instead. (The parent node will not be deleted by anyone in this case though)

The implementation with a "raw pointer" or even a reference is potentially dangerous, because the controller is stored in a "shared pointer" inside a controller node (parent node). So if the controller (`pure_pursuit` or `mpc`) is copied somewhere outside of the controller node (parent node) - it can outlive the controller node (parent node) and try to access the invalid memory of the already destroyed node.

This is why I proposed a solution independent from the node. The controller copies only the necessary dependencies from the node and does not hold any reference to if whatsoever. The node reference parameter is used only to initialize the dependencies (clock, logger etc.) and is passed to other components for initialization only (mostly for parameter declarations). 
This means that the node reference parameter is used only for the initialization - no node reference is being held by the controller after the construction. This means a controller can safely outlive the parent node.

TakaHoribe reviewed (approved):
@TauTheLepton thank you. Your explanation totally make sense. Please merge this PR if you are ok.

TauTheLepton commented:
@TakaHoribe 

> @TauTheLepton thank you. Your explanation totally make sense. Please merge this PR if you are ok.

I am okay with the changes and would like to merge the PR, but unfortunately I can't because I don't have write access to this repository.

Could you, or anybody with write access, merge this PR please.
